File: Front/src/types/quote.ts
/**
 * Types for quote management
 */

export type QuoteType = 'workshop' | 'product' | 'custom';
export type QuoteStatus = 'pending' | 'quoted' | 'approved' | 'rejected' | 'completed';
export type ClientType = 'particulier' | 'entreprise';

/**
 * Quote Request - Input for creating a quote
 */
export interface QuoteRequestIn {
  quote_type: QuoteType;
  title: string;
  description: string;
  quantity: number;
  client_type: ClientType;
  client_name: string;
  client_email: string;
  client_phone: string;
  company_name?: string;
}

/**
 * Quote Update - Input for updating a quote (admin only)
 */
export interface QuoteUpdateIn {
  final_price?: number;
  admin_notes?: string;
  artisan_id?: string;
}

/**
 * Quote - Full quote object returned from API
 */
export interface Quote {
  id: string;
  user_id: string;
  artisan_id?: string | null;
  quote_type: QuoteType;
  title: string;
  description: string;
  quantity: number;
  client_type: ClientType;
  client_name: string;
  client_email: string;
  client_phone: string;
  company_name?: string | null;
  status: QuoteStatus;
  estimated_price?: number | null;
  final_price?: number | null;
  admin_notes?: string | null;
  requested_at: string; // ISO datetime
  quoted_at?: string | null;
  responded_at?: string | null;
  completed_at?: string | null;
}

/**
 * Quote List Response - Paginated list of quotes
 */
export interface QuoteListResponse {
  items: Quote[];
  total: number;
  skip: number;
  limit: number;
  stats?: QuoteStats;
}

/**
 * Quote Stats - Statistics about quotes
 */
export interface QuoteStats {
  total: number;
  pending: number;
  quoted: number;
  approved: number;
  rejected: number;
  completed: number;
  approval_rate?: number; // percentage (0-100)
  conversion_rate?: number; // percentage (0-100)
  average_response_time?: number; // hours
  total_value?: number;
}

/**
 * Quote Summary Response - When creating/updating a quote
 */
export interface QuoteSummaryResponse {
  id: string;
  status: QuoteStatus;
  message: string;
  final_price?: number;
}

/**
 * Quote Conversion Response - When converting quote to order
 */
export interface QuoteConversionResponse {
  order_id: string;
  order_data: {
    title: string;
    description: string;
    quantity: number;
    price: number;
    status: string;
  };
  message: string;
}

File: Back/app/models/quote.py
"""Quote model for custom quotes management"""
from sqlalchemy import Column, String, Numeric, ForeignKey, DateTime
from sqlalchemy import Integer, Text
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
import uuid
from app.models.base import BaseModel, GUID


class Quote(BaseModel):
    """
    Quote records for custom product/workshop quotes.
    Allows users to request custom quotes for artisans.
    Admin can manage quotes and convert to orders.
    """
    __tablename__ = "quotes"

    id = Column(GUID, primary_key=True, default=uuid.uuid4, index=True)

    # Users involved
    user_id = Column(
        GUID,
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True
    )
    artisan_id = Column(
        GUID,
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
        index=True
    )

    # Quote details
    quote_type = Column(String(50), nullable=False)  # workshop/product/custom
    title = Column(String(255), nullable=False)
    description = Column(Text, nullable=False)
    quantity = Column(Integer, nullable=False, default=1)

    # Client information
    client_type = Column(String(50), nullable=False)  # particulier/entreprise
    client_name = Column(String(255), nullable=False)
    client_email = Column(String(255), nullable=False)
    client_phone = Column(String(20), nullable=False)
    company_name = Column(String(255), nullable=True)

    # Quote status and pricing
    status = Column(String(50), nullable=False, default="pending", index=True)
    # pending/quoted/approved/rejected/completed
    estimated_price = Column(Numeric(10, 2), nullable=True)
    final_price = Column(Numeric(10, 2), nullable=True)
    admin_notes = Column(Text, nullable=True)

    # Timeline
    requested_at = Column(
        DateTime,
        server_default=func.now(),
        nullable=False,
        index=True
    )
    quoted_at = Column(DateTime, nullable=True)
    responded_at = Column(DateTime, nullable=True)
    completed_at = Column(DateTime, nullable=True)
    created_at = Column(DateTime, server_default=func.now(), nullable=False)
    updated_at = Column(
        DateTime,
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False
    )

    # Relationships
    user = relationship(
        "User",
        foreign_keys=[user_id],
        backref="quotes_as_requester"
    )
    artisan = relationship(
        "User",
        foreign_keys=[artisan_id],
        backref="quotes_as_provider"
    )

    def __repr__(self):
        return f"<Quote {self.id} - {self.status} - {self.title}>"

